### PR TITLE
Complete Rework

### PR DESCRIPTION
**Pull request by derekShaheen**
Hey guys, I saw your youtube channel and really thought it was a cool project. I put aside some time and reworked each of the files in the project, but I don't have the hardware to actually test any of this, so it could be off. Anyway, if it's off here or there, I think this is a lot more readable, and should be a lot easier to extend or change the functionality. 

Please give it a review and maybe a try and let me know what you think.

### CODE REVIEW

**Comment by derekShaheen** [comment]
dish_scan.py
1. Removing redundant print statements.
2. Creating a function input_with_range_check to handle input with range checks, thus reducing code repetition.
3. Using f-strings to concatenate strings when saving the settings.
4. Creating a function send_dish_command to handle sending commands to the dish, thus reducing code repetition.
5. Removing unnecessary print statements and using f-strings for a cleaner output.
6. Using list comprehensions and the bytes function to simplify the code when sending characters to the dish.

fix_image.py
1. I created functions load_data, extract_timestamp, process_data, and plot_heatmap to modularize the code.
2. I simplified the data processing for both resolution 1 and 2.
3. I removed the unnecessary print statements and used f-strings for cleaner output.
4. I added a main guard to allow the script to be imported as a module without running the main code block.

dish_image.py
1. I reused the functions load_data, extract_timestamp, and plot_heatmap from fix_image.py.
2. I created a new process_data function that trims the messy edges of the array and adjusts the motor/indexing issue with the dish and scan code.
3. I simplified the data processing by slicing the numpy array and using a for loop to shift every other row of the matrix.
4. I removed the unnecessary print statements and used f-strings for cleaner output.
5. I added a main guard to allow the script to be imported as a module without running the main code block.

**Review by jaagupku — APPROVED** [review]
looks good

**Comment by saveitforparts** [comment]
I'm still learning how GitHub works and puzzling through the first pull request I got :-P

**Comment by derekShaheen** [comment]
Hey there Gabe, no worries, it can be complicated. I reviewed the other two pull requests, and my changes won't be compatible with either of these. It appears you have the following decision to make. If you take no action, then you've selected option 1.

1. You want to keep your current code, perhaps only implementing change #1 (or none if you don't want to).
2. You want to use #2, then you'll probably want to accept #1 and #2.
3. You want to use this, then simply close the other two for now and accept this one.

It looks like #2 has a lot of improvements as well. Unfortunately I don't have time to review those changes fully, however. I might recommend trying both #2 and this one and see what you think about them from there. Perhaps that other user will have some time to review and may input on this as well.

**Comment by RubenKelevra** [comment]
Hey @derekShaheen,

cool to see someone else having interest in improving this project :)

Hey @saveitforparts,

you sound a bit overwhelmed. How about creating an org instead and put the code there? This way, you could give others the ability to help you maintain this project together with you?

https://github.com/organizations/new

**Comment by RubenKelevra** [comment]
@derekShaheen wrote:
> It looks like https://github.com/saveitforparts/Tailgater-Microwave-Imaging/pull/2 has a lot of improvements as well. Unfortunately I don't have time to review those changes fully, however. I might recommend trying both https://github.com/saveitforparts/Tailgater-Microwave-Imaging/pull/2 and this one and see what you think about them from there. Perhaps that other user will have some time to review and may input on this as well.

No worries. Maybe we can break down the changes we want to do in smaller chunks?

Main improvement in my changes would probably be, to use the nudging feature better:

It just nudges 4 times, and then does absolute positioning to align the antenna correctly again.

This needs to be checked with the hardware. It's maybe necessary to first move the antenna back to a different position and then send the new absolute position to avoid backlash in the gearbox.

**Comment by saveitforparts** [comment]
> you sound a bit overwhelmed. How about creating an org instead and put the code there? This way, you could give others the ability to help you maintain this project together with you?

I have to admit I got over my head with this Github thing. I've used the website before as a place to snag interesting code for small projects, but I don't actually use git on a regular basis. I figured this was a good spot to dump my code, vs an obscure zip file on my obscure website. I just wasn't prepared for the amount of interest it generated! I also have a bunch of other projects I'm in the middle of, so I'm not necessarily working on this one every day. I also don't fully understand some of the code and methods you guys are suggesting, so I'm not 100% sure how to fix errors or make changes.

Since there is a bunch of interest, I'll try to keep on top of this, but my list of potential code updates might not be exactly what other people are thinking. For example, I might need to drop the "high res" / nudge idea altogether since the beamwidth of the antenna is >1 degree and it won't physically "see" any smaller than the low res version. However, one of the original devs for the hardware (the guy who wrote the "nudge" function) just contacted me, so maybe he can tell me more. 

I did briefly try out PR #1 and #2 and had some issues, but didn't have time to really dig into them. I haven't had time to try your #3 yet. Since each full run of the dish takes about 3.5 hours plus some setup, it's something I have to schedule ahead of time. 

TL:DR I'm a Python / GitHub noob but I'll keep poking at it :-)
